test/common/http/http2/codec_impl_test.cc

```diff
@@ -612,6 +612,35 @@ TEST_P(Http2CodecImplTest, BadMetadataVecReceivedTest) {
   EXPECT_THROW_WITH_MESSAGE(request_encoder_->encodeMetadata(metadata_map_vector), EnvoyException,
                             "The user callback function failed");
 }
+
+// Encode response metadata while dispatching request data from the client, so
+// that nghttp2 can't fill the metadata frames' payloads until dispatching
+// is finished.
+TEST_P(Http2CodecImplTest, EncodeMetadataWhileDispatchingTest) {
```

Comment:
Should we keep the old test and add the new test, or is this strictly speaking more realistic?
If the order of operations wasn't realistic in the prior unit tests I'm inclined to suggest we add an integration test as well.

Reply:
Both orders are realistic, depending on (I think) what kind of network event happens to be driving us.